[clang] implement current direction of CWG2765 for string literal comparisons in constant evaluation

zygoloid opened the pull request:
Track the identity of each string literal object produced by evaluation
with a global version number. Accept comparisons between literals of the
same version, and between literals of different versions that cannot
possibly be placed in overlapping storage. Treat the remaining
comparisons as non-constant.

zygoloid committed:
Implement current CWG direction for string literal comparisons.

Track the identity of each string literal object produced by evaluation
with a global version number. Accept comparisons between literals of the
same version, and between literals of different versions that cannot
possibly be placed in overlapping storage. Treat the remaining
comparisons as non-constant.

zygoloid committed:
Add a test that repeated evaluation in the same function call might
produce different string literals each time.

zygoloid committed:
A couple more tests.

shafik reviewed (commented):
In case I don't get a chance to look at this in more detail right away can you please reference specific CWG issues in the summary and in comments in the code? Thank you.

mizvekov commented on clang/test/SemaCXX/constant-expression-cxx14.cpp:
```diff
@@ -1306,3 +1306,18 @@ constexpr int field(int a) {
+
+namespace literal_comparison {
+
+constexpr bool different_in_loop(bool b = false) {
+  if (b) return false;
```
What's the purpose of this 'b' parameter?

zygoloid replied:
Just to suppress the "constexpr function never produces a constant expression" diagnostic.

tbaederr commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2061,15 +2063,21 @@ static bool EvaluateIgnoredValue(EvalInfo &Info, const Expr *E) {
 }
 
+static bool IsOpaqueConstantCall(const LValue &LVal) {
+  auto *BaseExpr =
+      llvm::dyn_cast_or_null<CallExpr>(LVal.Base.dyn_cast<const Expr *>());
```
```suggestion
  const auto *BaseExpr =
      llvm::dyn_cast_if_present<CallExpr>(LVal.Base.dyn_cast<const Expr *>());
```

tbaederr commented on clang/include/clang/Basic/DiagnosticASTKinds.td:
```diff
@@ -96,6 +96,8 @@ def note_constexpr_pointer_constant_comparison : Note<
 def note_constexpr_literal_comparison : Note<
   "comparison of addresses of literals has unspecified value">;
+def note_constexpr_opaque_call_comparison : Note<
+  "comparison against opaque constant has unspecified value">;
```
I can't imagine a normal person knows what's meant by "opaque constant"

AaronBallman replied:
Yeah, opaque constant is pretty opaque, but I'm struggling to think of a better way to phrase this. It only triggers when calling a builtin which forms an address constant.

Mayyyybbbe: comparison against an anonymous object with a constant address has unspecified value

? (or maybe 'unnamed' instead of 'anonymous')

zygoloid replied:
It's not necessarily anonymous, or an object. For example, we get one of these constants from `__builtin_start_address(function)`, where the returned address is often the address of the function (though in some cases it might be a different code address -- and we don't know).

Added the actual pointer value to the diagnostic:
```
note: comparison against opaque constant address '&__builtin_function_start(a)' has unspecified value
```
Hopefully that makes it a bit clearer what the problem is :)

tbaederr commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,81 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+
+// Determine whether two string literals potentially overlap. This will be the
+// case if they agree on the values of all the bytes on the overlapping region
+// between them.
```
"The overlapping region" means they share some prefix?

zygoloid replied:
If LHS is `"abcdef" + 3` and RHS is `"cdef\0gh" + 1` then the overlapping region is `cdef\0` -- the portion of the two string literals that must overlap in memory if the pointers actually point to the same address at runtime.

Would it be helpful to add something like that to the comment?

erichkeane replied:
I think that sort of example would be incredibly helpful next time I go through this.

zygoloid replied:
Added.

AaronBallman commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,81 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+  // within RHS. We don't need to look at the characters of one string that
+  // would appear before the start of the other string if they were merged.
+  CharUnits Offset = RHS.Offset - LHS.Offset;
+  if (Offset.isPositive()) {
+    RHSString.Bytes = RHSString.Bytes.drop_front(Offset.getQuantity());
+  } else if (Offset.isNegative()) {
+    LHSString.Bytes = LHSString.Bytes.drop_front(-Offset.getQuantity());
+  }
```
```suggestion
  if (Offset.isPositive())
    RHSString.Bytes = RHSString.Bytes.drop_front(Offset.getQuantity());
  else if (Offset.isNegative())
    LHSString.Bytes = LHSString.Bytes.drop_front(-Offset.getQuantity());
```

AaronBallman commented on clang/test/SemaCXX/constant-expression-cxx14.cpp:
```diff
@@ -1306,3 +1306,18 @@ constexpr int field(int a) {
+constexpr bool check = different_in_loop();
+  // expected-error@-1 {{}} expected-note@-1 {{in call}}
+
+}
```
Please add a newline to the end of the file.

AaronBallman commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2061,15 +2063,21 @@ static bool EvaluateIgnoredValue(EvalInfo &Info, const Expr *E) {
+/// Should this call expression be treated as forming an opaque constant?
+static bool IsOpaqueConstantCall(const CallExpr *E) {
   unsigned Builtin = E->getBuiltinCallee();
   return (Builtin == Builtin::BI__builtin___CFStringMakeConstantString ||
           Builtin == Builtin::BI__builtin___NSStringMakeConstantString ||
           Builtin == Builtin::BI__builtin_ptrauth_sign_constant ||
```
Can we get test coverage for these cases?

zygoloid replied:
Added.

AaronBallman commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -324,6 +324,13 @@ class ASTContext : public RefCountedBase<ASTContext> {
+  /// This is used to distinguish between repeated evaluations of the same
+  /// string literal.
+  ///
+  /// TODO: Ensure version numbers don't collide when deserialized.
```
Should we add a test case involving modules to make sure that behavior is reasonable?

AaronBallman reviewed (commented):
Thank you for working on this! Be sure to update the release notes so users know about the fix.

erichkeane commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -324,6 +324,13 @@ class ASTContext : public RefCountedBase<ASTContext> {
+  /// This is used to distinguish between repeated evaluations of the same
+  /// string literal.
+  ///
+  /// TODO: Ensure version numbers don't collide when deserialized.
```
This seems troublesome.  The modules case here is somewhat concerning to me.  I wonder if we could figure out some way to encode/serialize the "how we got here" via source information to provide the number?  I realize I'm being a little hand-wavy, but perhaps back to the 'line + col that caused this evaluation'?

zygoloid replied:
It's not too hard to contrive a situation where there's an observable misbehavior. For example:

```c++
export module A;
export constexpr const char *f() { return "hello"; }
```
```c++
export module B;
import A;
export constexpr const char *p = f();
```
```c++
export module C;
import A;
export constexpr const char *q = f();
```
```c++
import B;
import C;
// Should be non-constant, but these both have version 0 and
// the same `StringLiteral`, so presumably passes.
static_assert(p == q);
// Might return false at runtime in practice.
bool g() { return p == q; }
```

It's a somewhat low-probability misbehavior at least -- you need the version numbers to collide across modules. (They can also collide with this patch after `unsigned` overflow, but I don't think we need to worry about that.)

Maybe this has too high a risk of collisions still, but we could initialize this counter to a hash of the current module name instead of 0. :)

If we want to do this properly, we could track the greatest version used in each module, and offset the versions in each module by the sum of the versions of earlier-loaded modules (and likewise offset the version for the current file by the sum across all modules). Alternatively we could try to recompute versions at import time, building a map of {module ID, imported version} -> new version. But it's not quite that simple, because given

```c++
export module A;
export constexpr const char *p = "foo";
```
```c++
export module B;
import A;
export constexpr const char *q = p;
```
we need the constant value of `p` imported from A to be the same as the constant value of `q` imported from B.

... So in the former case we would need to do proper remapping not just adding an offset, and in the latter case we'd need to track which module ID the imported version *originally* came from. Both options seem expensive.

Maybe we can think of a better way? (Initializing the counter to a hash is sounding a bit better now, huh?)

erichkeane replied:
> Maybe we can think of a better way? (Initializing the counter to a hash is sounding a bit better now, huh?)

Thats probably at least as good as my line/column/file info suggestion, but based on the above/thinking further, I think anything besides "start at the same number for all" is perfectly acceptable.

erichkeane commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,81 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+  CharUnits Offset = RHS.Offset - LHS.Offset;
+  if (Offset.isPositive()) {
+    RHSString.Bytes = RHSString.Bytes.drop_front(Offset.getQuantity());
+  } else if (Offset.isNegative()) {
```
I'd probably drop the 'isPositive' branch/check and make this just:

```
if (Offset.isNegative()))
  RHSString.Bytes = RHSString.Bytes.drop_front(-Offset.getQuantity());
else
  RHSString.Bytes = RHSString.Bytes.drop_front(Offset.getQuantity());
```
As 'drop_front' of 0 is well defined, right?


Though, this just ends up being this, right?
```
  RHSString.Bytes = RHSString.Bytes.drop_front(std::abs(Offset.getQuantity()));
```

zygoloid replied:
Sure, we can `drop_front(0)`; removed the `isPositive` check. Done.

Note that one branch is removing characters from the start of `LHSString` and the other is removing characters from the start of `RHSString`, so the suggested rewrites aren't correct.

erichkeane replied:
OH!  I missed that re `LHSString` vs `RHSString`. Thanks, and definitely explains why you did it that way, thanks.

erichkeane commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,81 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+  bool LHSIsLonger = LHSString.Bytes.size() > RHSString.Bytes.size();
+  StringRef Longer = LHSIsLonger ? LHSString.Bytes : RHSString.Bytes;
+  StringRef Shorter = LHSIsLonger ? RHSString.Bytes : LHSString.Bytes;
+  int ShorterCharWidth = (LHSIsLonger ? RHSString : LHSString).CharWidth;
```
Is it a problem/issue if the `CharWidth` of the two is different?  If they're stored with different widths, can they really be overlapping?

zygoloid replied:
I don't see why they couldn't overlap in that case. There's an example of this in the tests -- if we have both a string literal `"A\0\0"` and a string literal `u"A"` on a little-endian target, the byte representation for both is `{65, 0, 0, 0}`, so I don't see a reason why an implementation wouldn't be permitted to use the same storage for both. The rule in [intro.object]/10.3 doesn't consider the types of the objects.

In practice, [Clang doesn't combine the strings in that case, but GCC does](https://godbolt.org/z/bd81vTYe7).

erichkeane replied:
Ah, neat!  I hadn't considered that actual memory representation would be combined-on, rather than the strings itself.  I question the value of this collapse (as it requires arbitrarily/strangely small values), but possible all the same.  Thanks for the clarification.

erichkeane reviewed (commented):
Needs a release note

zygoloid committed:
Use _if_present not _or_null

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

zygoloid committed:
Remove redundant braces.

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

zygoloid committed:
Improve diagnosis for opaque constant calls.

zygoloid committed:
Address some review comments.

zygoloid committed:
Add release note.

zygoloid committed:
Testing for other opaque constant calls.

Tweak diagnostic to avoid suggesting this is a rule from a language standard.

zygoloid committed:
Add a comment mentioning the CWG issue.

zygoloid reviewed (commented):
We need to figure out what we want to do about modules, but I think I've addressed the rest of the comments.

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -74,6 +74,22 @@ C++ Specific Potentially Breaking Changes
+- During constant evaluation, comparisons between different evaluations of the
+  same string literal are now correctly treated as non-constant, and comparisons
+  between string literals that cannot possibly overlap in memory are now treated
+  as constant.
```
Could you mention the CWG issue here as well?

zygoloid replied:
Done.

zygoloid committed:
Mention core issue in release note.

erichkeane reviewed (approved):
I'm happy once SOME level of solution for the index-starting-value is done (`NextStringLiteralVersion=rand()` 🤡).  I think the avoided collisions is worth a touch of effort, but not much  more.  

I think there is some concern from others about `opaque` that should probably be settled on, but I'm happy with whatever all of you come up with, even if it is `opaque`.

zygoloid committed:
Add test for modules behavior.

zygoloid committed:
Extend test and fix comment.

zygoloid commented:
> I'm happy once SOME level of solution for the index-starting-value is done (`NextStringLiteralVersion=rand()` 🤡). I think the avoided collisions is worth a touch of effort, but not much more.

Um. So. I added a test for this before I started working on a fix, and ... the test passes. It turns out that the existing implementation already works: while we do have some support for importing computed constant values, we always recompute constants locally before using them as inputs to further local constant evaluations, so we produce a local numbering even for what appear to be imported constants.

We would need to do something different here if we ever change that, but we now have test coverage for it, so we should notice.

> I think there is some concern from others about `opaque` that should probably be settled on, but I'm happy with whatever all of you come up with, even if it is `opaque`.

I hope I've addressed that by including the value of the pointer in the diagnostic.

efriedma-quic commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -8573,7 +8661,10 @@ class LValueExprEvaluator
   bool VisitMemberExpr(const MemberExpr *E);
-  bool VisitStringLiteral(const StringLiteral *E) { return Success(E); }
+  bool VisitStringLiteral(const StringLiteral *E) {
+    uint64_t Version = Info.getASTContext().getNextStringLiteralVersion();
```
getNextStringLiteralVersion() returns unsigned, not uint64_t?

zygoloid replied:
Fixed. (Sorry, just some junk left behind from a previous approach.)

zygoloid committed:
Remove leftover handling of uint64_t

tbaederr commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,91 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+  // The null terminator isn't included in the string data, so check for it
+  // manually. If the longer string doesn't have a null terminator where the
+  // shorter string ends, they aren't potentially overlapping.
+  for (int nullByte : llvm::seq(ShorterCharWidth)) {
```
```suggestion
  for (int NullByte : llvm::seq(ShorterCharWidth)) {
```

tbaederr commented on clang/test/AST/ByteCode/cxx20.cpp:
```diff
@@ -108,22 +108,16 @@ constexpr auto p2 = "test2";
+constexpr auto b3 = name1() == name1(); // ref-error {{must be initialized by a constant expression}} \
+                                        // ref-note {{comparison of addresses of literals}}
+constexpr auto b4 = name1() == name2();
+static_assert(!b4);
```
Just from looking at these few lines, I don't understand why `b3` warns but `b4` doesn't. They both compare the address of literals.

The bytecode interpreter simply creates global variables for string literals, so `b3` here is simply true, like it was in the current interpreter before. Is this behavior wrong now?

zygoloid replied:
The C++ model in general is that each evaluation of a string literal expression can produce a distinct object, and that these objects can fully or partially overlap each other in memory when they have suitable values; that's the model that this PR is implementing.

`b3` is non-constant because each call to `name1()` can produce a distinct value, so the comparison result is unspecified. `b4` is constant and `false` under this PR because the values returned by `name1()` and `name2()` cannot possibly be the same -- those two string literal evaluations can't produce the same value because the strings have different contents.

zygoloid commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2142,11 +2150,91 @@ static const ValueDecl *GetLValueBaseDecl(const LValue &LVal) {
+  // manually. If the longer string doesn't have a null terminator where the
+  // shorter string ends, they aren't potentially overlapping.
+  for (int nullByte : llvm::seq(ShorterCharWidth)) {
+    if (Shorter.size() + nullByte >= Longer.size())
+      break;
+    if (Longer[Shorter.size() + nullByte])
```
```suggestion
    if (Shorter.size() + NullByte >= Longer.size())
      break;
    if (Longer[Shorter.size() + NullByte])
```

zygoloid committed:
Fix capitalization.

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

zygoloid commented:
> I worry about string literals in vague-linkage entities, because the "version" of the string literal which is referred to from an inline-function/variable at runtime may not actually be the same version of the string literal seen in the current compilation -- at runtime we may in fact be using a different version of the symbol from another TU.

I think string literals in inline functions shouldn't be a problem by themselves -- each evaluation of the string literal is allowed to produce a different object, and each constant evaluation will produce a different version. But I think there is a problem in general for string literals that appear in the evaluated value of a vague linkage constant. For example:
```c++
constexpr inline const char *s = "foo";
```
... is required by the standard to have the same value across all translation units, but we currently don't ensure that's the case. I think that's a separable issue from this PR, but it's clearly related -- though I think probably the right answer here is for the ABI to assign a mangling to the string literal in this case. See https://github.com/itanium-cxx-abi/cxx-abi/issues/78 and https://github.com/llvm/llvm-project/issues/57957.

jyknight commented:
> I think that's a separable issue from this PR

Yep. I brought it up mainly in case consideration of that set of issues might necessitate design changes here. It sounds like that's not the case though, so I'm happy for it to not be a blocker for this PR.

zygoloid commented:
I believe all the feedback has been addressed, and plan to merge in 24 hours if there are no further comments.

zeroomega commented:
Hi,

We are seeing build error on fmtlib project after this patch is landed. The error message from the clang is:

```
FAILED: host_x64/obj/third_party/fmtlib/src/src/fmtlib.os.cc.o 
 ../../prebuilt/third_party/clang/custom/bin/clang++ -MD -MF host_x64/obj/third_party/fmtlib/src/src/fmtlib.os.cc.o.d -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_REMOVE_TRANSITIVE_INCLUDES -I../.. -Ihost_x64/gen -I../../third_party/fmtlib/src/include -fcolor-diagnostics -fcrash-diagnostics-dir=clang-crashreports -fcrash-diagnostics=all -gen-reproducer=error -ffp-contract=off --sysroot=../../prebuilt/third_party/sysroot/linux --target=x86_64-unknown-linux-gnu -ffile-compilation-dir=. -no-canonical-prefixes -fomit-frame-pointer -fdata-sections -ffunction-sections -O0 -Xclang -debug-info-kind=constructor -g3 -grecord-gcc-switches -gdwarf-5 -gz=zstd -Wall -Wextra -Wconversion -Wextra-semi -Wimplicit-fallthrough -Wnewline-eof -Wstrict-prototypes -Wwrite-strings -Wno-sign-conversion -Wno-unused-parameter -Wnonportable-system-include-path -Wno-missing-field-initializers -Wno-extra-qualification -Wno-cast-function-type-strict -Wno-cast-function-type-mismatch -Wno-unknown-warning-option -Wno-missing-template-arg-list-after-template-kw -Wno-deprecated-pragma -fvisibility=hidden -Werror -Wa,--fatal-warnings -Wno-error=deprecated-declarations --sysroot=../../prebuilt/third_party/sysroot/linux --target=x86_64-unknown-linux-gnu -fPIE -fvisibility-inlines-hidden -stdlib=libc++ -stdlib=libc++ -std=c++20 -Wno-deprecated-this-capture -fno-exceptions -fno-rtti -ftemplate-backtrace-limit=0 -stdlib=libc++ -c ../../third_party/fmtlib/src/src/os.cc -o host_x64/obj/third_party/fmtlib/src/src/fmtlib.os.cc.o
 ../../third_party/fmtlib/src/src/os.cc:176:35: error: call to consteval function 'fmt::basic_format_string<char, const char *>::basic_format_string<FMT_COMPILE_STRING, 0>' is not a constant expression
   176 |     FMT_THROW(system_error(errno, FMT_STRING("cannot open file {}"),
       |                                   ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/base.h:779:54: note: subexpression not valid in a constant expression
   779 |     format_str_.remove_prefix(detail::to_unsigned(it - begin()));
       |                                                   ~~~^~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2769:5: note: in call to 'this->context_.advance_to(&"cannot open file {}"[18])'
  2769 |     context_.advance_to(begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2764:5: note: in call to 'this->on_format_specs(0, &"cannot open file {}"[18], &"cannot open file {}"[18])'
  2764 |     on_format_specs(id, begin, begin);  // Call parse() on empty specs.
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2571:5: note: in call to 'handler.on_replacement_field(0, &"cannot open file {}"[18])'
  2571 |     handler.on_replacement_field(handler.on_arg_id(), begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2621:21: note: in call to 'parse_replacement_field<char, fmt::detail::format_string_checker<char, const char *> &>(&"cannot open file {}"[18], &"cannot open file {}"[19], checker(s))'
  2621 |         begin = p = parse_replacement_field(p - 1, end, handler);
       |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2892:7: note: in call to 'parse_format_string<true, char, fmt::detail::format_string_checker<char, const char *>>({&"cannot open file {}"[0], 19}, checker(s))'
  2892 |       detail::parse_format_string<true>(str_, checker(s));
       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/src/os.cc:176:35: note: in call to 'basic_format_string<FMT_COMPILE_STRING, 0>([] {
     struct __attribute__((visibility("hidden")))  FMT_COMPILE_STRING : fmt::detail::compile_string {
         using char_type [[maybe_unused]] = fmt::remove_cvref_t<decltype("cannot open file {}"[0])>;
         [[maybe_unused]] constexpr operator fmt::basic_string_view<char_type>() const {
             return fmt::detail_exported::compile_string_to_view<char_type>("cannot open file {}");
         }
     };
     return FMT_COMPILE_STRING();
 }())'
   176 |     FMT_THROW(system_error(errno, FMT_STRING("cannot open file {}"),
       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   177 |                            filename.c_str()));
       |                            ~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/format.h:136:55: note: expanded from macro 'FMT_THROW'
   136 |       ::fmt::detail::assert_fail(__FILE__, __LINE__, (x).what())
       |                                                       ^
 ../../third_party/fmtlib/src/src/os.cc:222:29: error: call to consteval function 'fmt::basic_format_string<char, const char *>::basic_format_string<FMT_COMPILE_STRING, 0>' is not a constant expression
   222 |         system_error(errno, FMT_STRING("cannot open file {}"), path.c_str()));
       |                             ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/base.h:779:54: note: subexpression not valid in a constant expression
   779 |     format_str_.remove_prefix(detail::to_unsigned(it - begin()));
       |                                                   ~~~^~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2769:5: note: in call to 'this->context_.advance_to(&"cannot open file {}"[18])'
  2769 |     context_.advance_to(begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2764:5: note: in call to 'this->on_format_specs(0, &"cannot open file {}"[18], &"cannot open file {}"[18])'
  2764 |     on_format_specs(id, begin, begin);  // Call parse() on empty specs.
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2571:5: note: in call to 'handler.on_replacement_field(0, &"cannot open file {}"[18])'
  2571 |     handler.on_replacement_field(handler.on_arg_id(), begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2621:21: note: in call to 'parse_replacement_field<char, fmt::detail::format_string_checker<char, const char *> &>(&"cannot open file {}"[18], &"cannot open file {}"[19], checker(s))'
  2621 |         begin = p = parse_replacement_field(p - 1, end, handler);
       |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2892:7: note: in call to 'parse_format_string<true, char, fmt::detail::format_string_checker<char, const char *>>({&"cannot open file {}"[0], 19}, checker(s))'
  2892 |       detail::parse_format_string<true>(str_, checker(s));
       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/src/os.cc:222:29: note: in call to 'basic_format_string<FMT_COMPILE_STRING, 0>([] {
     struct __attribute__((visibility("hidden")))  FMT_COMPILE_STRING : fmt::detail::compile_string {
         using char_type [[maybe_unused]] = fmt::remove_cvref_t<decltype("cannot open file {}"[0])>;
         [[maybe_unused]] constexpr operator fmt::basic_string_view<char_type>() const {
             return fmt::detail_exported::compile_string_to_view<char_type>("cannot open file {}");
         }
     };
     return FMT_COMPILE_STRING();
 }())'
   221 |     FMT_THROW(
       |     ~~~~~~~~~~
   222 |         system_error(errno, FMT_STRING("cannot open file {}"), path.c_str()));
       |         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/format.h:136:55: note: expanded from macro 'FMT_THROW'
   136 |       ::fmt::detail::assert_fail(__FILE__, __LINE__, (x).what())
       |                                                       ^
 ../../third_party/fmtlib/src/src/os.cc:291:16: error: call to consteval function 'fmt::basic_format_string<char, int &>::basic_format_string<FMT_COMPILE_STRING, 0>' is not a constant expression
   291 |         errno, FMT_STRING("cannot duplicate file descriptor {}"), fd));
       |                ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/base.h:779:54: note: subexpression not valid in a constant expression
   779 |     format_str_.remove_prefix(detail::to_unsigned(it - begin()));
       |                                                   ~~~^~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2769:5: note: in call to 'this->context_.advance_to(&"cannot duplicate file descriptor {}"[34])'
  2769 |     context_.advance_to(begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2764:5: note: in call to 'this->on_format_specs(0, &"cannot duplicate file descriptor {}"[34], &"cannot duplicate file descriptor {}"[34])'
  2764 |     on_format_specs(id, begin, begin);  // Call parse() on empty specs.
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2571:5: note: in call to 'handler.on_replacement_field(0, &"cannot duplicate file descriptor {}"[34])'
  2571 |     handler.on_replacement_field(handler.on_arg_id(), begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2655:13: note: in call to 'parse_replacement_field<char, fmt::detail::format_string_checker<char, int> &>(&"cannot duplicate file descriptor {}"[34], &"cannot duplicate file descriptor {}"[35], checker(s))'
  2655 |     begin = parse_replacement_field(p, end, handler);
       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2892:7: note: in call to 'parse_format_string<true, char, fmt::detail::format_string_checker<char, int>>({&"cannot duplicate file descriptor {}"[0], 35}, checker(s))'
  2892 |       detail::parse_format_string<true>(str_, checker(s));
       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/src/os.cc:291:16: note: in call to 'basic_format_string<FMT_COMPILE_STRING, 0>([] {
     struct __attribute__((visibility("hidden")))  FMT_COMPILE_STRING : fmt::detail::compile_string {
         using char_type [[maybe_unused]] = fmt::remove_cvref_t<decltype("cannot duplicate file descriptor {}"[0])>;
         [[maybe_unused]] constexpr operator fmt::basic_string_view<char_type>() const {
             return fmt::detail_exported::compile_string_to_view<char_type>("cannot duplicate file descriptor {}");
         }
     };
     return FMT_COMPILE_STRING();
 }())'
   290 |     FMT_THROW(system_error(
       |     ~~~~~~~~~~~~~~~~~~~~~~~
   291 |         errno, FMT_STRING("cannot duplicate file descriptor {}"), fd));
       |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/format.h:136:55: note: expanded from macro 'FMT_THROW'
   136 |       ::fmt::detail::assert_fail(__FILE__, __LINE__, (x).what())
       |                                                       ^
 ../../third_party/fmtlib/src/src/os.cc:300:16: error: call to consteval function 'fmt::basic_format_string<char, int &, int &>::basic_format_string<FMT_COMPILE_STRING, 0>' is not a constant expression
   300 |         errno, FMT_STRING("cannot duplicate file descriptor {} to {}"), fd_,
       |                ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/base.h:779:54: note: subexpression not valid in a constant expression
   779 |     format_str_.remove_prefix(detail::to_unsigned(it - begin()));
       |                                                   ~~~^~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2769:5: note: in call to 'this->context_.advance_to(&"cannot duplicate file descriptor {} to {}"[34])'
  2769 |     context_.advance_to(begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2764:5: note: in call to 'this->on_format_specs(0, &"cannot duplicate file descriptor {} to {}"[34], &"cannot duplicate file descriptor {} to {}"[34])'
  2764 |     on_format_specs(id, begin, begin);  // Call parse() on empty specs.
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2571:5: note: in call to 'handler.on_replacement_field(0, &"cannot duplicate file descriptor {} to {}"[34])'
  2571 |     handler.on_replacement_field(handler.on_arg_id(), begin);
       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2655:13: note: in call to 'parse_replacement_field<char, fmt::detail::format_string_checker<char, int, int> &>(&"cannot duplicate file descriptor {} to {}"[34], &"cannot duplicate file descriptor {} to {}"[41], checker(s))'
  2655 |     begin = parse_replacement_field(p, end, handler);
       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/include/fmt/base.h:2892:7: note: in call to 'parse_format_string<true, char, fmt::detail::format_string_checker<char, int, int>>({&"cannot duplicate file descriptor {} to {}"[0], 41}, checker(s))'
  2892 |       detail::parse_format_string<true>(str_, checker(s));
       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ../../third_party/fmtlib/src/src/os.cc:300:16: note: in call to 'basic_format_string<FMT_COMPILE_STRING, 0>([] {
     struct __attribute__((visibility("hidden")))  FMT_COMPILE_STRING : fmt::detail::compile_string {
         using char_type [[maybe_unused]] = fmt::remove_cvref_t<decltype("cannot duplicate file descriptor {} to {}"[0])>;
         [[maybe_unused]] constexpr operator fmt::basic_string_view<char_type>() const {
             return fmt::detail_exported::compile_string_to_view<char_type>("cannot duplicate file descriptor {} to {}");
         }
     };
     return FMT_COMPILE_STRING();
 }())'
   299 |     FMT_THROW(system_error(
       |     ~~~~~~~~~~~~~~~~~~~~~~~
   300 |         errno, FMT_STRING("cannot duplicate file descriptor {} to {}"), fd_,
       |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   301 |         fd));
       |         ~~~~
 ../../third_party/fmtlib/src/include/fmt/format.h:1842:23: note: expanded from macro 'FMT_STRING'
  1842 | #define FMT_STRING(s) FMT_STRING_IMPL(s, fmt::detail::compile_string, )
       |                       ^
 ../../third_party/fmtlib/src/include/fmt/format.h:1821:3: note: expanded from macro 'FMT_STRING_IMPL'
  1821 |   [] {                                                                        \
       |   ^
 ../../third_party/fmtlib/src/include/fmt/format.h:136:55: note: expanded from macro 'FMT_THROW'
   136 |       ::fmt::detail::assert_fail(__FILE__, __LINE__, (x).what())
       |                                                       ^
 4 errors generated.
```

Full build log can be seen from: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8735688601552461713/+/u/build/ninja/stdout

We are still looking into it to understand if it is a compiler bug or an issue with fmtlib's implementation.

zygoloid commented:
That backtrace doesn't match the fmtlib code available from their official repository. Where can I find the code you're building?

zygoloid commented:
I found this: https://fuchsia.googlesource.com/third_party/github.com/fmtlib/fmt/+/refs/heads/upstream/main/include/fmt/base.h#2664

Note that this is passing `s` separately both to `parse_format_string` and to `checker`. Here, `s` is a lambda that implicitly converts to a string view by evaluating a string literal. Those two evaluations can return different string literal objects, and I'm assuming pointers to those strings are what we end up subtracting.

So yes, this fmtlib code is wrong -- it's incorrectly assuming that a string literal will always evaluate to the same value, which is not true at runtime or (after this Clang change) at compile time. The correct thing to do would be what is done a few lines below in the next constructor: convert the `S` object to a string view once, and then use that same value for both calls.

zygoloid commented:
I've filed a fmtlib bug: https://github.com/fmtlib/fmt/issues/4177
